### PR TITLE
virsh cpu and mem fixes

### PR DESCRIPTION
**Pull request by jfehlig**
### PR creator: Description

Similar to PR#935, improve the virsh CPU and memory sections of the VT guide. Add notes about special configuration needed for large CPU and memory VMs.

### PR creator: Are there any relevant issues/feature requests?

* bsc#1188751

### PR creator: Which product versions do the changes apply to?

When opening a PR, check all versions of the documentation that your PR applies to.

- SLE 15/openSUSE Leap 15.x
  - [x] SLE 15 SP4/openSUSE Leap 15.4 *(current `main`, no backport necessary)*
  - [x] SLE 15 SP3/openSUSE Leap 15.3
  - [ ] SLE 15 SP2/openSUSE Leap 15.2
  - [ ] SLE 15 SP1
  - [ ] SLE 15 GA
- SLE 12
  - [ ] SLE 12 SP5
  - [ ] SLE 12 SP4
  - [ ] SLE 12 SP3

### PR reviewer only: Have all backports been applied?

The doc team member merging your PR will take care of backporting to older documents.
When opening a PR, do *not* set the following check box.

- [ ] all necessary backports are done

### CODE REVIEW

File: xml/libvirt_configuration_virsh.xml [hunk 1]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    The CPU model exposed to a &vmguest; can often influence the workload
+    running within it. The default CPU model is derived from a cpu mode
+    known as host-model.
+   </para>
+<screen>&lt;cpu mode='host-model'/&gt;</screen>
+   <para>
+    When starting a &vmguest; with cpu mode host-model, &libvirt; will
+    copy its model of the host CPU into the &vmguest; definition. The host
+    CPU model and features copied to the &vmguest; definition can be
+    observed in the output of 'virsh capabilities'.
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:340]:
```suggestion
    observed in the output of the <command>virsh capabilities</command>.
```

File: xml/libvirt_configuration_virsh.xml [hunk 2]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+   </para>
+   <para>
+    Another interesting CPU mode is host-passthrough.
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:343]:
```suggestion
    Another interesting CPU mode is <literal>host-passthrough</literal>.
```

File: xml/libvirt_configuration_virsh.xml [hunk 3]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+   </para>
+<screen>&lt;cpu mode='host-passthrough'/&gt;</screen>
+   <para>
+    When starting a &vmguest; with CPU mode host-passthrough, it is presented
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:347]:
```suggestion
    When starting a &vmguest; with CPU mode <literal>host-passthrough</literal>, it is presented
```

File: xml/libvirt_configuration_virsh.xml [hunk 4]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    with a CPU that is exactly the same as the &vmhost; CPU. This can be
+    useful when the &vmguest; workload requires CPU features not available in
+    &libvirt;'s simplified host-model CPU. The host-passthrough CPU mode is
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:350]:
```suggestion
    &libvirt;'s simplified <literal>host-model</literal> CPU. The <literal>host-passthrough</literal> CPU mode is
```

File: xml/libvirt_configuration_virsh.xml [hunk 5]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    also required in some cases, e.g. when running &vmguest;s with more than
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:351]:
```suggestion
    also required in some cases, for example, when running &vmguest;s with more than
```

File: xml/libvirt_configuration_virsh.xml [hunk 6]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    4TB of memory. The host-passthrough CPU mode comes with the disadvantage of
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:352]:
```suggestion
    4TB of memory. The <literal>host-passthrough</literal> CPU mode comes with the disadvantage of
```

File: xml/libvirt_configuration_virsh.xml [hunk 7]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    reduced migration flexibility. A &vmguest; with host-passthrough CPU mode
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:353]:
```suggestion
    reduced migration flexibility. A &vmguest; with <literal>host-passthrough</literal> CPU mode
```

File: xml/libvirt_configuration_virsh.xml [hunk 8]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    can only be migrated to a &vmhost; with identical hardware.
+   </para>
+   <para>
+    When using the host-passthrough CPU mode, it is still possible to
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:357]:
```suggestion
    When using the <literal>host-passthrough</literal> CPU mode, it is still possible to
```

File: xml/libvirt_configuration_virsh.xml [hunk 9]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+    disable undesirable features. The following configuration will present
+    the &vmguest; with a CPU that is exactly the same as the host CPU but
+    with the vmx feature disabled.
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:360]:
```suggestion
    with the <literal>vmx</literal> feature disabled.
```

File: xml/libvirt_configuration_virsh.xml [hunk 10]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+   </para>
+<screen>&lt;cpu mode='host-passthrough'&gt;
+   &lt;feature policy='disable' name='vmx'/&gt;
+   &lt;/cpu&gt;
+</screen>
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:365]:
```suggestion
<screen>
&lt;cpu mode='host-passthrough'&gt;
  &lt;feature policy='disable' name='vmx'/&gt;
  &lt;/cpu&gt;
</screen>
```

File: xml/libvirt_configuration_virsh.xml [hunk 11]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+   <para>
+    The custom CPU mode is another common mode used to define a normalized
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:367]:
```suggestion
    The <literal>custom</literal> CPU mode is another common mode used to define a normalized
```

File: xml/libvirt_configuration_virsh.xml [hunk 12]
```diff
@@ -257,9 +266,138 @@ CPU:            N/A
...
+  &lt;feature policy='require' name='vmx'/&gt;
+  &lt;feature policy='require' name='est'/&gt;
+  &lt;feature policy='require' name='tm2'/&gt;
+  &lt;feature policy='require' name='xtpr'/&gt;
+  &lt;feature policy='require' name='pdcm'/&gt;
+  &lt;feature policy='require' name='dca'/&gt;
+  &lt;feature policy='require' name='rdtscp'/&gt;
+  &lt;feature policy='require' name='invtsc'/&gt;
+  &lt;/cpu&gt;
+</screen>
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:394]:
```suggestion
<screen>
&lt;cpu mode='custom' match='exact'&gt;
  &lt;model fallback='allow'&gt;Nehalem&lt;/model&gt;
  &lt;feature policy='require' name='vme'/&gt;
  &lt;feature policy='require' name='ds'/&gt;
  &lt;feature policy='require' name='acpi'/&gt;
  &lt;feature policy='require' name='ss'/&gt;
  &lt;feature policy='require' name='ht'/&gt;
  &lt;feature policy='require' name='tm'/&gt;
  &lt;feature policy='require' name='pbe'/&gt;
  &lt;feature policy='require' name='dtes64'/&gt;
  &lt;feature policy='require' name='monitor'/&gt;
  &lt;feature policy='require' name='ds_cpl'/&gt;
  &lt;feature policy='require' name='vmx'/&gt;
  &lt;feature policy='require' name='est'/&gt;
  &lt;feature policy='require' name='tm2'/&gt;
  &lt;feature policy='require' name='xtpr'/&gt;
  &lt;feature policy='require' name='pdcm'/&gt;
  &lt;feature policy='require' name='dca'/&gt;
  &lt;feature policy='require' name='rdtscp'/&gt;
  &lt;feature policy='require' name='invtsc'/&gt;
  &lt;/cpu&gt;
</screen>
```

File: xml/libvirt_configuration_virsh.xml [hunk 13]
```diff
@@ -376,7 +514,19 @@ CPU Affinity:   yy
 
   <para>
    The amount of memory allocated for the &vmguest; can also be configured with
-   virsh. It is stored in the <tag>memory</tag> element. Follow these steps:
+   virsh. It is stored in the <tag>memory</tag> element and defines the
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:517]:
```suggestion
   &virsh;. It is stored in the <tag>memory</tag> element and defines the
```

File: xml/libvirt_configuration_virsh.xml [hunk 14]
```diff
@@ -376,7 +514,19 @@ CPU Affinity:   yy
...
+   maximum allocation of memory for the &vmguest; at boot time. The optional
+   <tag>currentMemory</tag> element defines the actual memory allocated to
+   the &vmguest;. <tag>currentMemory</tag> can be less than <tag>memory</tag>,
+   allowing for increasing, or ballooning, the memory while the &vmguest; is
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:521]:
```suggestion
   allowing for increasing (or <emphasis>ballooning</emphasiss> the memory while the &vmguest; is
```

File: xml/libvirt_configuration_virsh.xml [hunk 15]
```diff
@@ -376,7 +514,19 @@ CPU Affinity:   yy
...
+   running. If <tag>currentMemory</tag> is omitted, it defaults to the same
+   value as the <tag>memory</tag> element.
+  </para>
+  <para>
+   Memory settings can be adjusted by editing the &vmguest; configuration,
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:526]:
```suggestion
   You can adjust memory settings by editing the &vmguest; configuration,
```

File: xml/libvirt_configuration_virsh.xml [hunk 16]
```diff
@@ -388,20 +538,62 @@ CPU Affinity:   yy
...
    <step>
     <para>
-     Check the amount of allocated RAM in your VM by running:
+     If the <tag>currentMemory</tag> element does not exist, add it below
+     the <tag>memory</tag> element, or change its value to 4G:
+    </para>
+<screen>...
+&lt;memory unit='KiB'&gt;8388608&lt;/memory&gt;
+&lt;currentMemory unit='KiB'&gt;4194304&lt;/currentMemory&gt;
+...</screen>
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:555]:
```suggestion
<screen>
[...]
&lt;memory unit='KiB'&gt;8388608&lt;/memory&gt;
&lt;currentMemory unit='KiB'&gt;4194304&lt;/currentMemory&gt;
[...]</screen>
```

File: xml/libvirt_configuration_virsh.xml [hunk 17]
```diff
@@ -388,20 +538,62 @@ CPU Affinity:   yy
...
    <step>
     <para>
-     Search for the <tag>memory</tag> element and set the amount of allocated
-     RAM:
+     Search for the <tag>memory</tag> element and set to 8G:
     </para>
 <screen>...
-&lt;memory unit='KiB'&gt;524288&lt;/memory&gt;
+&lt;memory unit='KiB'&gt;8388608&lt;/memory&gt;
 ...</screen>
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:545]:
```
<screen>
[...]
&lt;memory unit='KiB'&gt;8388608&lt;/memory&gt;
[...]
</screen>
```

File: xml/libvirt_configuration_virsh.xml [hunk 18]
```diff
@@ -388,20 +538,62 @@ CPU Affinity:   yy
...
+    </para>
+<screen>...
+&lt;memory unit='KiB'&gt;8388608&lt;/memory&gt;
+&lt;currentMemory unit='KiB'&gt;4194304&lt;/currentMemory&gt;
+...</screen>
+   </step>
+  </procedure>
+  <para>
+   Changing the memory allocation while the &vmguest; is running can be done
+   with the setmem subcommand. The following example shows increasing the
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:560]:
```suggestion
   with the <command>setmem</command> subcommand. The following example shows increasing the
```

File: xml/libvirt_configuration_virsh.xml [hunk 19]
```diff
@@ -388,20 +538,62 @@ CPU Affinity:   yy
...
+   memory allocation to 8G:
+  </para>
+
+   <procedure>
+   <step>
+    <para>
+     Check &vmguest; existing memory settings
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:567]:
```suggestion
     Check &vmguest; existing memory settings:
```

File: xml/libvirt_configuration_virsh.xml [hunk 20]
```diff
@@ -388,20 +538,62 @@ CPU Affinity:   yy
...
+   <step>
+    <para>
+     Change the used memory to 8G:
     </para>
-<screen>&prompt.user;<command>cat /proc/meminfo</command></screen>
+<screen>&prompt.sudo;<command>virsh setmem sles15 8388608</command></screen>
+   </step>
+   <step>
+    <para>
+     Check the updated memory settings
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:582]:
```suggestion
     Check the updated memory settings:
```

File: xml/libvirt_configuration_virsh.xml [hunk 21]
```diff
@@ -388,20 +538,62 @@ CPU Affinity:   yy
...
+Max memory:     8388608 KiB
+Used memory:    8388608 KiB
+</screen>
    </step>
   </procedure>
+  <important>
+   <title>Large memory &vmguest;s</title>
+   <para>
+    &vmguests; with memory requirements of 4TB or more must currently
+    use the host-passthrough CPU model.
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:594]:
```suggestion
    use the <literal>host-passthrough</literal> CPU model.
```

**Review by tbazant — COMMENTED** [review]
Great work, minor updates suggested

File: xml/libvirt_configuration_virsh.xml (hunk 14 above)
Comment (jfehlig) [reply xml/libvirt_configuration_virsh.xml:521]: This change doesn't look quite right. Probably just the closing ')' is missing, e.g.
(or <emphasis>ballooning</emphasiss>)
I don't know how to fix it as part of this review process, so will not "Add suggestion to batch" for this one.

**Review by tbazant — APPROVED** [review]
LGTM

**Pull request merged**